File: sdks/ts/memory/src/ingest/extract/keyframe.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Keyframe extraction from video files. Captures frames at a
 * configurable interval using FFmpeg, then runs OCR on each frame
 * via a delegate extractor. Results are appended to the video
 * extraction output.
 */

import { mkdir, readFile, readdir } from 'node:fs/promises'
import { join } from 'node:path'

import { checkBinaryAvailable, runSubprocess } from './subprocess.js'
import type { ExtractOptions, ExtractResult, Extractor } from './types.js'

/** Default interval between keyframe captures in seconds. */
const DEFAULT_KEYFRAME_INTERVAL_SECS = 30

/** Configuration for keyframe extraction. */
export type KeyframeConfig = {
  readonly ffmpegBinary: string
  readonly intervalSecs: number
  readonly maxKeyframes: number
  readonly ocrExtractor?: Extractor
  readonly timeout: number
}

/** Text extracted from a single keyframe. */
export type KeyframeResult = {
  readonly timestampSecs: number
  readonly text: string
  readonly confidence: number
}

/** Resolves keyframe config with defaults and env var overrides. */
export const resolveKeyframeConfig = (partial: Partial<KeyframeConfig>): KeyframeConfig => {
  const envInterval = process.env['MEMORY_KEYFRAME_INTERVAL_SECS']
  let intervalSecs = partial.intervalSecs ?? DEFAULT_KEYFRAME_INTERVAL_SECS
  if (partial.intervalSecs === undefined && envInterval !== undefined && envInterval !== '') {
    const parsed = Number.parseInt(envInterval, 10)
    if (!Number.isNaN(parsed) && parsed > 0) {
      intervalSecs = parsed
    }
  }

  const base: KeyframeConfig = {
    ffmpegBinary: partial.ffmpegBinary ?? process.env['MEMORY_FFMPEG_PATH'] ?? 'ffmpeg',
    intervalSecs,
    maxKeyframes: partial.maxKeyframes ?? 20,
    timeout: partial.timeout ?? 60_000,
  }
  if (partial.ocrExtractor !== undefined) {
    return { ...base, ocrExtractor: partial.ocrExtractor }
  }
  return base
}

/**
 * Extracts keyframe images from a video file at the configured
 * interval, runs OCR on each frame, and returns the collected text
 * results.
 */
export const extractKeyframes = async (
  videoPath: string,
  tmpDir: string,
  cfg: KeyframeConfig,
  signal?: AbortSignal,
): Promise<readonly KeyframeResult[]> => {
  if (cfg.ocrExtractor === undefined) {
    return []
  }

  const ocrAvailable = await cfg.ocrExtractor.available()
  if (!ocrAvailable) {
    return []
  }

  const framesDir = join(tmpDir, 'keyframes')
  await mkdir(framesDir, { recursive: true })

  const fpsFilter = `fps=1/${cfg.intervalSecs}`
  const outputPattern = join(framesDir, 'frame-%04d.png')

  const args = [
    '-i', videoPath,
    '-vf', fpsFilter,
    '-frames:v', String(cfg.maxKeyframes),
    '-vsync', 'vfr',
    outputPattern,
  ]

  const subprocessOpts = signal !== undefined
    ? { timeout: cfg.timeout, signal }
    : { timeout: cfg.timeout }

  const result = await runSubprocess(cfg.ffmpegBinary, args, undefined, subprocessOpts)
  if (result.exitCode !== 0) {
    throw new Error(
      `ingest: ffmpeg keyframe extraction exited with code ${result.exitCode}: ${result.stderr}`,
    )
  }

  const files = await readdir(framesDir)
  let frameFiles = files
    .filter((f) => f.startsWith('frame-') && f.endsWith('.png'))
    .sort()

  if (frameFiles.length > cfg.maxKeyframes) {
    frameFiles = frameFiles.slice(0, cfg.maxKeyframes)
  }

  const keyframeResults: KeyframeResult[] = []

  for (let i = 0; i < frameFiles.length; i++) {
    const frameFile = frameFiles[i]
    if (frameFile === undefined) continue

    const framePath = join(framesDir, frameFile)

    let imageData: Buffer
    try {
      imageData = await readFile(framePath)
    } catch {
      continue
    }

    const ocrOpts: ExtractOptions = {
      contentType: 'image/png',
      fileName: frameFile,
    }

    try {
      const ocrResult = await cfg.ocrExtractor.extract(imageData, ocrOpts, signal)
      if (!ocrResult.skipped && ocrResult.text.length > 0) {
        const timestampSecs = (i + 1) * cfg.intervalSecs

        keyframeResults.push({
          timestampSecs,
          text: ocrResult.text,
          confidence: ocrResult.confidence,
        })
      }
    } catch {
      continue
    }
  }

  return keyframeResults
}

/**
 * Formats keyframe OCR results into a structured text section suitable
 * for appending to the video extraction result.
 */
export const formatKeyframeText = (results: readonly KeyframeResult[]): string => {
  if (results.length === 0) return ''

  const lines: string[] = ['\n\n--- Keyframe Text ---']

  for (const kf of results) {
    const timestamp = formatKeyframeTimestamp(kf.timestampSecs)
    lines.push(`\n[${timestamp}]\n${kf.text}`)
  }

  return lines.join('\n')
}

/**
 * Converts seconds to MM:SS format.
 */
export const formatKeyframeTimestamp = (seconds: number): string => {
  const mins = Math.floor(seconds / 60)
  const secs = Math.floor(seconds % 60)
  return `${String(mins).padStart(2, '0')}:${String(secs).padStart(2, '0')}`
}

/**
 * Adds keyframe-related metadata entries to the provided metadata
 * record. Returns a new record with the additions.
 */
export const mergeKeyframeMetadata = (
  metadata: Record<string, string>,
  results: readonly KeyframeResult[],
): void => {
  metadata['keyframe_count'] = String(results.length)

  if (results.length === 0) return

  let totalConf = 0
  for (const kf of results) {
    totalConf += kf.confidence
  }
  const avgConf = totalConf / results.length
  metadata['keyframe_avg_confidence'] = avgConf.toFixed(4)
}

File: go/ingest/keyframe.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

// defaultKeyframeInterval is the default time between keyframe
// captures in seconds. Override via MEMORY_KEYFRAME_INTERVAL_SECS.
const defaultKeyframeInterval = 30

// KeyframeConfig holds the resolved keyframe extraction settings.
type KeyframeConfig struct {
	FFmpegBinary string
	IntervalSecs int
	MaxKeyframes int
	OCRExtractor Extractor
	Timeout      time.Duration
}

// KeyframeResult holds the text extracted from a single keyframe.
type KeyframeResult struct {
	TimestampSecs float64
	Text          string
	Confidence    float64
}

// applyKeyframeDefaults fills zero-value fields with defaults.
func applyKeyframeDefaults(cfg *KeyframeConfig) {
	if cfg.FFmpegBinary == "" {
		cfg.FFmpegBinary = envOrDefault("MEMORY_FFMPEG_PATH", "ffmpeg")
	}
	if cfg.IntervalSecs <= 0 {
		if envVal := os.Getenv("MEMORY_KEYFRAME_INTERVAL_SECS"); envVal != "" {
			if v, err := strconv.Atoi(envVal); err == nil && v > 0 {
				cfg.IntervalSecs = v
			}
		}
		if cfg.IntervalSecs <= 0 {
			cfg.IntervalSecs = defaultKeyframeInterval
		}
	}
	if cfg.MaxKeyframes <= 0 {
		cfg.MaxKeyframes = defaultMaxKeyframes
	}
	if cfg.Timeout <= 0 {
		cfg.Timeout = 60 * time.Second
	}
}

// extractKeyframes extracts keyframe images from a video file at the
// configured interval, runs OCR on each frame, and returns the
// collected text results. The caller is responsible for the tmpDir
// lifecycle.
func extractKeyframes(ctx context.Context, videoPath, tmpDir string, cfg KeyframeConfig) ([]KeyframeResult, error) {
	applyKeyframeDefaults(&cfg)

	if cfg.OCRExtractor == nil {
		return nil, nil
	}

	available, err := cfg.OCRExtractor.Available(ctx)
	if err != nil || !available {
		return nil, nil
	}

	framesDir := filepath.Join(tmpDir, "keyframes")
	if err := os.MkdirAll(framesDir, 0o700); err != nil {
		return nil, fmt.Errorf("ingest: creating keyframes dir: %w", err)
	}

	// Extract keyframes using FFmpeg fps filter.
	fpsFilter := fmt.Sprintf("fps=1/%d", cfg.IntervalSecs)
	outputPattern := filepath.Join(framesDir, "frame-%04d.png")

	args := []string{
		"-i", videoPath,
		"-vf", fpsFilter,
		"-frames:v", strconv.Itoa(cfg.MaxKeyframes),
		"-vsync", "vfr",
		outputPattern,
	}

	result, err := RunSubprocess(ctx, cfg.FFmpegBinary, args, nil, SubprocessOptions{
		Timeout: cfg.Timeout,
	})
	if err != nil {
		return nil, fmt.Errorf("ingest: keyframe extraction: %w", err)
	}
	if result.ExitCode != 0 {
		return nil, fmt.Errorf("ingest: ffmpeg keyframe extraction exited with code %d: %s",
			result.ExitCode, result.Stderr)
	}

	// Collect generated frame images.
	frameFiles, err := filepath.Glob(filepath.Join(framesDir, "frame-*.png"))
	if err != nil {
		return nil, fmt.Errorf("ingest: listing keyframe images: %w", err)
	}

	sort.Strings(frameFiles)

	// Cap at MaxKeyframes.
	if len(frameFiles) > cfg.MaxKeyframes {
		frameFiles = frameFiles[:cfg.MaxKeyframes]
	}

	var keyframeResults []KeyframeResult

	for i, framePath := range frameFiles {
		frameData, readErr := os.ReadFile(framePath)
		if readErr != nil {
			continue
		}

		ocrOpts := ExtractOptions{
			ContentType: "image/png",
			FileName:    filepath.Base(framePath),
		}

		ocrResult, ocrErr := cfg.OCRExtractor.Extract(ctx, frameData, ocrOpts)
		if ocrErr != nil || ocrResult.Skipped || ocrResult.Text == "" {
			continue
		}

		timestampSecs := float64(i+1) * float64(cfg.IntervalSecs)

		keyframeResults = append(keyframeResults, KeyframeResult{
			TimestampSecs: timestampSecs,
			Text:          ocrResult.Text,
			Confidence:    ocrResult.Confidence,
		})
	}

	return keyframeResults, nil
}

// formatKeyframeText formats keyframe OCR results into a structured
// text section suitable for appending to the video extraction result.
func formatKeyframeText(results []KeyframeResult) string {
	if len(results) == 0 {
		return ""
	}

	var sb strings.Builder
	sb.WriteString("\n\n--- Keyframe Text ---\n")

	for _, kf := range results {
		timestamp := formatKeyframeTimestamp(kf.TimestampSecs)
		sb.WriteString(fmt.Sprintf("\n[%s]\n%s\n", timestamp, kf.Text))
	}

	return sb.String()
}

// formatKeyframeTimestamp converts seconds to MM:SS format.
func formatKeyframeTimestamp(seconds float64) string {
	mins := int(seconds) / 60
	secs := int(seconds) % 60
	return fmt.Sprintf("%02d:%02d", mins, secs)
}

// mergeKeyframeMetadata adds keyframe-related metadata entries to the
// provided metadata map. Mutates the map in place.
func mergeKeyframeMetadata(metadata map[string]string, results []KeyframeResult) {
	metadata["keyframe_count"] = strconv.Itoa(len(results))

	if len(results) == 0 {
		return
	}

	var totalConf float64
	for _, kf := range results {
		totalConf += kf.Confidence
	}
	avgConf := totalConf / float64(len(results))
	metadata["keyframe_avg_confidence"] = formatConfidence(avgConf)
}
